File: golang/01_soap_ingles/soap_client.go
package main

import (
    "bytes"
    "fmt"
    "io"
    "net/http"
    "regexp"
    "strings"
)

func consultarSOAP(n int) (string, error) {
    xml := fmt.Sprintf(`<?xml version="1.0" encoding="utf-8"?>
<soap:Envelope xmlns:soap="http://schemas.xmlsoap.org/soap/envelope/">
  <soap:Body>
    <NumberToWords xmlns="http://www.dataaccess.com/webservicesserver/">
      <ubiNum>%d</ubiNum>
    </NumberToWords>
  </soap:Body>
</soap:Envelope>`, n)

    req, err := http.NewRequest("POST", "https://www.dataaccess.com/webservicesserver/NumberConversion.wso", bytes.NewBufferString(xml))
    if err != nil { return "", err }
    req.Header.Set("Content-Type", "text/xml; charset=utf-8")
    req.Header.Set("SOAPAction", "http://www.dataaccess.com/webservicesserver/NumberToWords")

    resp, err := http.DefaultClient.Do(req)
    if err != nil { return "", err }
    defer resp.Body.Close()

    body, _ := io.ReadAll(resp.Body)
    re := regexp.MustCompile(`<m:NumberToWordsResult>(.*?)</m:NumberToWordsResult>|<NumberToWordsResult>(.*?)</NumberToWordsResult>`)
    match := re.FindStringSubmatch(string(body))
    if len(match) >= 3 {
        if match[1] != "" { return strings.TrimSpace(match[1]), nil }
        return strings.TrimSpace(match[2]), nil
    }
    return "Resultado no encontrado", nil
}

File: rust/02_soap_traducido/src/soap_client.rs
pub fn consultar_soap(n: i32) -> Result<String, Box<dyn std::error::Error>> {
    let xml = format!(r#"<?xml version="1.0" encoding="utf-8"?>
<soap:Envelope xmlns:soap="http://schemas.xmlsoap.org/soap/envelope/">
  <soap:Body>
    <NumberToWords xmlns="http://www.dataaccess.com/webservicesserver/">
      <ubiNum>{}</ubiNum>
    </NumberToWords>
  </soap:Body>
</soap:Envelope>"#, n);

    let body = reqwest::blocking::Client::new()
        .post("https://www.dataaccess.com/webservicesserver/NumberConversion.wso")
        .header("Content-Type", "text/xml; charset=utf-8")
        .header("SOAPAction", "http://www.dataaccess.com/webservicesserver/NumberToWords")
        .body(xml)
        .send()?
        .text()?;

    let re = regex::Regex::new(r"<m:NumberToWordsResult>(.*?)</m:NumberToWordsResult>|<NumberToWordsResult>(.*?)</NumberToWordsResult>")?;
    if let Some(caps) = re.captures(&body) {
        if let Some(m) = caps.get(1) { return Ok(m.as_str().trim().to_string()); }
        if let Some(m) = caps.get(2) { return Ok(m.as_str().trim().to_string()); }
    }
    Ok("Resultado no encontrado".to_string())
}
